enhancements/mcd-reboot-exceptions.md

```diff
@@ -0,0 +1,247 @@
+   ordered list of known actions that are needed to apply it
+1. Modify any parts of MCD that assume drain and reboot always happens after
+   writing to disk
+1. Modify MCD to perform only the list of actions produced in step 1
+
+The full list of post-write actions required for this feature to be useful are:
+
+- drain
+- reboot
+- restart a named systemd service
+
+If the action results in an error, or takes too long, a reboot will be
+performed as in prior 4.x versions.
+
+### User Stories [optional]
+
+The user stories below are intended to be illustrative of the different types of 
+services we need to build capabilities for.
+
+#### Story 1
+
+As an administrator of a disconnected bare metal cluster, I want the ability to
+update the image catalog mirror[1] without potentially causing a full cluster
+reboot, so that I can obtain the latest fixes quickly and without needing to be
+concerned about the cumulative disruption to the cluster.
+
+[1] As well as the ICSP associated with it, which is sensitive to the addition,
+subtraction, or renaming of images.
+
+#### Story 2
+
+As an adminstrator of a large bare metal cluster, I want the addition or removal
+of SSH keys to happen without a full cluster reboot, so that the credentials are
+usable (or revoked) in a timeframe comparable with traditional environments not
+hours.
+
+#### Story 3
+
+As an adminstrator of a bare metal cluster that takes a day to reboot, I want
+the system to use the least invasive method for applying configuration changes,
+so that I have periods of relative stability where I can accomodate the
+maintenance windows of less enlightened vendors.
+
+#### Story 4
+
+As an admin, I want to continue to force reboots for network configuration 
+updates and changes to kernel flags, so that they are applied in a timely 
+manner.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+How the comparision determines what constitutes a domain, whether by target file
+or by configuration element, is left for later discussion.
+
+### Risks and Mitigations
+
+The risks are of false positives (an update that requires a reboot does not
+trigger one), and false negatives (an update that does not require a reboot
+triggers one anyway).  While the latter may be unexpected by an admin that knows
+of this feature, at worst it is a performance issue and cannot be considered a
+regression over the current behaviour.
+
+False negatives can be mitigated by defaulting to the current behaviour
+(reboot), preventing the admin from tampering with the list of domains with
+exceptions, and only adding new domains after Engineering and QE validate that
+the changes are safe.
+
+## Design Details
+
+### Test Plan
+
+- Unit tests for the comparision function that cover
+  - Identical, missing, and invalid MachineConfigs
+  - Changes to domains for which no exception exists
+  - Changes to only domains for which a 'do nothing' exception exists
```

Comment:
There's no previous explanation of "exceptions"

Reply:
Added an explanation: the case when unit of configuration can be applied by a means other than the default drain+reboot behaviour